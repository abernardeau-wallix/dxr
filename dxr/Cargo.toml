[package]
name = "dxr"
description = "Declarative XML-RPC"
license = "MIT OR Apache-2.0"

version = "0.4.0-dev"
edition = "2021"
rust-version = "1.56.0"

authors = ["Fabio Valentini <decathorpe@gmail.com>"]
categories = ["network-programming", "web-programming::http-client", "encoding"]
keywords = ["XML-RPC", "client", "derive"]
repository = "https://github.com/ironthree/dxr"
exclude = ["/wip/"]

[[example]]
name = "client"
path = "examples/client.rs"
required-features = ["client"]

[[example]]
name = "koji"
path = "examples/koji.rs"
required-features = ["client", "derive", "nil"]

[[example]]
name = "server"
path = "examples/server.rs"
required-features = ["axum-server"]

[[test]]
name = "adder"
path = "tests/adder.rs"
required-features = ["client", "axum-server", "tokio"]

[[test]]
name = "derive"
path = "tests/derive.rs"
required-features = ["derive", "i8", "nil"]

[[test]]
name = "echo_any"
path = "tests/echo_any.rs"
required-features = ["derive", "client", "axum-server", "i8", "nil", "tokio"]

[[test]]
name = "echo_one"
path = "tests/echo_one.rs"
required-features = ["derive", "client", "axum-server", "i8", "nil", "tokio"]


[dependencies]
anyhow = "1.0.53"
base64 = "0.13"
chrono = "0.4.19"
log = "0.4.13"
quick-xml = { version = "0.22" , features = ["serialize"] }
serde = { version = "1.0.104", features = ["derive"] }
thiserror = "1.0.30"

# derive feature
dxr_derive = { path = "../dxr_derive", version = "0.4.0-dev", optional = true }

# client feature
reqwest = { version = "0.11", features = ["cookies"] , optional = true }
url = { version = "2.2", optional = true }

# server feature
async-trait = { version = "0.1.52", optional = true }
http = { version = "0.2.6", optional = true }

# server feature: axum support
axum = { version = "0.5.4", optional = true }
tokio = { version = "1.14",  optional = true }


[dev-dependencies]
# used for property-based conversion / roundrtip tests
quickcheck = "1.0"
quickcheck_macros = "1.0"

# used for example binaries and integration tests
tokio = { version = "1.14", features = ["macros", "rt-multi-thread", "time"] }

# used for compiling async doctests
tokio-test = "0.4.2"

# used for testing derive macros
trybuild = "1.0.56"


[features]
# basic client / server support
default = ["client", "server", "derive"]

# reqwest-based XML-RPC client support
client = ["reqwest", "http", "url"]

# basic XML-RPC server suport
server = ["http"]
# server implementation based on axum and tokio
axum-server = ["server", "async-trait", "axum", "tokio/time"]

# derive macros for ToDXR and FromDXR traits
derive = ["dxr_derive"]

# support non-standard <i8> XML-RPC values
i8 = []

# support non-standard <nil> XML-RPC values
nil = []


[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]
