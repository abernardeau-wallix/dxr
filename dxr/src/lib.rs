//! # dxr: declarative xml-rpc
//!
//! The `dxr` crate provides types, macros, and other functionality which can be used to write
//! fast and correct XML-API clients and servers in Rust conveniently.

#![warn(missing_docs)]
#![warn(missing_debug_implementations)]
#![warn(clippy::unwrap_used)]

#[cfg(feature = "derive")]
pub use dxr_derive::{FromDXR, ToDXR};

pub use dxr_shared::{DxrError, Fault, FromDXR, FromParams, ToDXR, ToParams, Value, XML_RPC_DATE_FORMAT};

// re-export chrono: DateTime / Utc are part of the public API
pub use dxr_shared::chrono;

// re-export url::Url: it is part of the public client API
#[cfg(feature = "client")]
pub use url;

// client-specific modules
#[cfg(feature = "client")]
mod call;
#[cfg(feature = "client")]
pub use call::*;

#[cfg(feature = "client")]
mod client;
#[cfg(feature = "client")]
pub use client::*;

// server-specific modules
#[cfg(feature = "server")]
mod server;
#[cfg(feature = "server")]
pub use server::*;

#[cfg(test)]
mod tests;
