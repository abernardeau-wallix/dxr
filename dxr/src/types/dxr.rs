mod from;
pub use from::*;

mod to;
pub use to::*;
