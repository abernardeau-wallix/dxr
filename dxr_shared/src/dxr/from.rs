use std::collections::HashMap;

use chrono::{DateTime, Utc};

use crate::error::DxrError;
use crate::traits::FromDXR;
use crate::types::{Type, Value};

impl FromDXR for Value {
    fn from_dxr(value: &Value) -> Result<Value, DxrError> {
        Ok(value.clone())
    }
}

impl FromDXR for i32 {
    fn from_dxr(value: &Value) -> Result<i32, DxrError> {
        match value.inner() {
            Type::Integer(int) => Ok(*int),
            t => Err(DxrError::wrong_type(t.name(), "i4")),
        }
    }
}

#[cfg(feature = "i8")]
impl FromDXR for i64 {
    fn from_dxr(value: &Value) -> Result<i64, DxrError> {
        match value.inner() {
            Type::Long(long) => Ok(*long),
            t => Err(DxrError::wrong_type(t.name(), "i8")),
        }
    }
}

impl FromDXR for bool {
    fn from_dxr(value: &Value) -> Result<bool, DxrError> {
        match value.inner() {
            Type::Boolean(boo) => Ok(*boo),
            t => Err(DxrError::wrong_type(t.name(), "boolean")),
        }
    }
}

impl FromDXR for String {
    fn from_dxr(value: &Value) -> Result<String, DxrError> {
        match value.inner() {
            Type::String(string) => Value::string_unescape(string),
            t => Err(DxrError::wrong_type(t.name(), "string")),
        }
    }
}

impl FromDXR for f64 {
    fn from_dxr(value: &Value) -> Result<f64, DxrError> {
        match value.inner() {
            Type::Double(double) => Ok(*double),
            t => Err(DxrError::wrong_type(t.name(), "double")),
        }
    }
}

impl FromDXR for DateTime<Utc> {
    fn from_dxr(value: &Value) -> Result<DateTime<Utc>, DxrError> {
        match value.inner() {
            Type::DateTime(date) => Ok(*date),
            t => Err(DxrError::wrong_type(t.name(), "dateTime.iso8861")),
        }
    }
}

impl FromDXR for Vec<u8> {
    fn from_dxr(value: &Value) -> Result<Vec<u8>, DxrError> {
        match value.inner() {
            Type::Base64(bytes) => Ok(bytes.clone()),
            t => Err(DxrError::wrong_type(t.name(), "base64")),
        }
    }
}

#[cfg(feature = "nil")]
impl<T> FromDXR for Option<T>
where
    T: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Option<T>, DxrError> {
        if let Type::Nil = value.inner() {
            Ok(None)
        } else {
            Ok(Some(T::from_dxr(value)?))
        }
    }
}

impl<T> FromDXR for Vec<T>
where
    T: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Vec<T>, DxrError> {
        let values = match value.inner() {
            Type::Array { data } => Ok(data.inner()),
            t => Err(DxrError::wrong_type(t.name(), "array")),
        };

        values?.iter().map(|value| T::from_dxr(value)).collect()
    }
}

impl<T> FromDXR for HashMap<String, T>
where
    T: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<HashMap<String, T>, DxrError> {
        let values = match value.inner() {
            Type::Struct { members } => Ok(members),
            t => Err(DxrError::wrong_type(t.name(), "struct")),
        };

        values?
            .iter()
            .map(|v| {
                let name = v.name().to_string();
                match T::from_dxr(v.inner()) {
                    Ok(value) => Ok((name, value)),
                    Err(error) => Err(error),
                }
            })
            .collect()
    }
}

// some implementations for exact numbers of values (with possibly different types)
impl FromDXR for () {
    fn from_dxr(value: &Value) -> Result<Self, DxrError> {
        match value.inner() {
            Type::Array { data } => {
                let values = data.inner();

                match values.len() {
                    0 => Ok(()),
                    n => Err(DxrError::return_mismatch(n, 0)),
                }
            },
            Type::Nil => Ok(()),
            other => Err(DxrError::wrong_type(other.name(), "array | nil")),
        }
    }
}

impl<T> FromDXR for (T,)
where
    T: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Self, DxrError> {
        if let Type::Array { data } = value.inner() {
            let values = data.inner();

            match values.len() {
                1 => {
                    let value = values.get(0).unwrap();

                    Ok((T::from_dxr(value)?,))
                },
                n => Err(DxrError::return_mismatch(n, 1)),
            }
        } else {
            Err(DxrError::wrong_type(value.inner().name(), "array"))
        }
    }
}

impl<A, B> FromDXR for (A, B)
where
    A: FromDXR,
    B: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Self, DxrError> {
        if let Type::Array { data } = value.inner() {
            let values = data.inner();

            match values.len() {
                2 => {
                    let a = values.get(0).unwrap();
                    let b = values.get(1).unwrap();

                    Ok((A::from_dxr(a)?, B::from_dxr(b)?))
                },
                n => Err(DxrError::return_mismatch(n, 2)),
            }
        } else {
            Err(DxrError::wrong_type(value.inner().name(), "array"))
        }
    }
}

impl<A, B, C> FromDXR for (A, B, C)
where
    A: FromDXR,
    B: FromDXR,
    C: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Self, DxrError> {
        if let Type::Array { data } = value.inner() {
            let values = data.inner();

            match values.len() {
                3 => {
                    let a = values.get(0).unwrap();
                    let b = values.get(1).unwrap();
                    let c = values.get(2).unwrap();

                    Ok((A::from_dxr(a)?, B::from_dxr(b)?, C::from_dxr(c)?))
                },
                n => Err(DxrError::return_mismatch(n, 3)),
            }
        } else {
            Err(DxrError::wrong_type(value.inner().name(), "array"))
        }
    }
}

impl<A, B, C, D> FromDXR for (A, B, C, D)
where
    A: FromDXR,
    B: FromDXR,
    C: FromDXR,
    D: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Self, DxrError> {
        if let Type::Array { data } = value.inner() {
            let values = data.inner();

            match values.len() {
                4 => {
                    let a = values.get(0).unwrap();
                    let b = values.get(1).unwrap();
                    let c = values.get(2).unwrap();
                    let d = values.get(3).unwrap();

                    Ok((A::from_dxr(a)?, B::from_dxr(b)?, C::from_dxr(c)?, D::from_dxr(d)?))
                },
                n => Err(DxrError::return_mismatch(n, 4)),
            }
        } else {
            Err(DxrError::wrong_type(value.inner().name(), "array"))
        }
    }
}

impl<A, B, C, D, E> FromDXR for (A, B, C, D, E)
where
    A: FromDXR,
    B: FromDXR,
    C: FromDXR,
    D: FromDXR,
    E: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Self, DxrError> {
        if let Type::Array { data } = value.inner() {
            let values = data.inner();

            match values.len() {
                5 => {
                    let a = values.get(0).unwrap();
                    let b = values.get(1).unwrap();
                    let c = values.get(2).unwrap();
                    let d = values.get(3).unwrap();
                    let e = values.get(4).unwrap();

                    Ok((
                        A::from_dxr(a)?,
                        B::from_dxr(b)?,
                        C::from_dxr(c)?,
                        D::from_dxr(d)?,
                        E::from_dxr(e)?,
                    ))
                },
                n => Err(DxrError::return_mismatch(n, 5)),
            }
        } else {
            Err(DxrError::wrong_type(value.inner().name(), "array"))
        }
    }
}

impl<A, B, C, D, E, F> FromDXR for (A, B, C, D, E, F)
where
    A: FromDXR,
    B: FromDXR,
    C: FromDXR,
    D: FromDXR,
    E: FromDXR,
    F: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Self, DxrError> {
        if let Type::Array { data } = value.inner() {
            let values = data.inner();

            match values.len() {
                6 => {
                    let a = values.get(0).unwrap();
                    let b = values.get(1).unwrap();
                    let c = values.get(2).unwrap();
                    let d = values.get(3).unwrap();
                    let e = values.get(4).unwrap();
                    let f = values.get(5).unwrap();

                    Ok((
                        A::from_dxr(a)?,
                        B::from_dxr(b)?,
                        C::from_dxr(c)?,
                        D::from_dxr(d)?,
                        E::from_dxr(e)?,
                        F::from_dxr(f)?,
                    ))
                },
                n => Err(DxrError::return_mismatch(n, 6)),
            }
        } else {
            Err(DxrError::wrong_type(value.inner().name(), "array"))
        }
    }
}

impl<A, B, C, D, E, F, G> FromDXR for (A, B, C, D, E, F, G)
where
    A: FromDXR,
    B: FromDXR,
    C: FromDXR,
    D: FromDXR,
    E: FromDXR,
    F: FromDXR,
    G: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Self, DxrError> {
        if let Type::Array { data } = value.inner() {
            let values = data.inner();

            match values.len() {
                7 => {
                    let a = values.get(0).unwrap();
                    let b = values.get(1).unwrap();
                    let c = values.get(2).unwrap();
                    let d = values.get(3).unwrap();
                    let e = values.get(4).unwrap();
                    let f = values.get(5).unwrap();
                    let g = values.get(6).unwrap();

                    Ok((
                        A::from_dxr(a)?,
                        B::from_dxr(b)?,
                        C::from_dxr(c)?,
                        D::from_dxr(d)?,
                        E::from_dxr(e)?,
                        F::from_dxr(f)?,
                        G::from_dxr(g)?,
                    ))
                },
                n => Err(DxrError::return_mismatch(n, 7)),
            }
        } else {
            Err(DxrError::wrong_type(value.inner().name(), "array"))
        }
    }
}

impl<A, B, C, D, E, F, G, H> FromDXR for (A, B, C, D, E, F, G, H)
where
    A: FromDXR,
    B: FromDXR,
    C: FromDXR,
    D: FromDXR,
    E: FromDXR,
    F: FromDXR,
    G: FromDXR,
    H: FromDXR,
{
    fn from_dxr(value: &Value) -> Result<Self, DxrError> {
        if let Type::Array { data } = value.inner() {
            let values = data.inner();

            match values.len() {
                8 => {
                    let a = values.get(0).unwrap();
                    let b = values.get(1).unwrap();
                    let c = values.get(2).unwrap();
                    let d = values.get(3).unwrap();
                    let e = values.get(4).unwrap();
                    let f = values.get(5).unwrap();
                    let g = values.get(6).unwrap();
                    let h = values.get(7).unwrap();

                    Ok((
                        A::from_dxr(a)?,
                        B::from_dxr(b)?,
                        C::from_dxr(c)?,
                        D::from_dxr(d)?,
                        E::from_dxr(e)?,
                        F::from_dxr(f)?,
                        G::from_dxr(g)?,
                        H::from_dxr(h)?,
                    ))
                },
                n => Err(DxrError::return_mismatch(n, 8)),
            }
        } else {
            Err(DxrError::wrong_type(value.inner().name(), "array"))
        }
    }
}

// if needed, implementations for more arguments can be implemented
