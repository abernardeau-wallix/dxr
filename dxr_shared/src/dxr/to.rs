use std::collections::HashMap;

use chrono::{DateTime, Utc};

use crate::error::DxrError;
use crate::traits::ToDXR;
use crate::types::{Array, Member, Struct, Value};
use crate::util::*;

impl ToDXR for Value {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(self.clone())
    }
}

impl ToDXR for &Value {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::clone(self))
    }
}

impl ToDXR for i32 {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::i4(*self))
    }
}

#[cfg(feature = "i8")]
#[cfg_attr(docsrs, doc(cfg(feature = "i8")))]
impl ToDXR for i64 {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::i8(*self))
    }
}

impl ToDXR for bool {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::boolean(*self))
    }
}

impl ToDXR for String {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        ToDXR::to_dxr(&self.as_str())
    }
}

impl ToDXR for &str {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Value::string_escape(self)
    }
}

impl ToDXR for f64 {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::double(*self))
    }
}

impl ToDXR for DateTime<Utc> {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::datetime(*self))
    }
}

impl ToDXR for Vec<u8> {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        ToDXR::to_dxr(&self.as_slice())
    }
}

impl<const N: usize> ToDXR for [u8; N] {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        ToDXR::to_dxr(&self.as_slice())
    }
}

impl ToDXR for &[u8] {
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::base64(self.to_vec()))
    }
}

#[cfg(feature = "nil")]
#[cfg_attr(docsrs, doc(cfg(feature = "nil")))]
impl<T> ToDXR for Option<T>
where
    T: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        if let Some(value) = self {
            T::to_dxr(value)
        } else {
            Ok(Value::nil())
        }
    }
}

#[cfg(feature = "nil")]
#[cfg_attr(docsrs, doc(cfg(feature = "nil")))]
impl<T> ToDXR for &Option<T>
where
    T: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        if let Some(value) = self {
            T::to_dxr(value)
        } else {
            Ok(Value::nil())
        }
    }
}

impl<T> ToDXR for Vec<T>
where
    T: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        ToDXR::to_dxr(&self.as_slice())
    }
}

impl<T, const N: usize> ToDXR for [T; N]
where
    T: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        ToDXR::to_dxr(&self.as_slice())
    }
}

impl<T> ToDXR for &[T]
where
    T: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        let values = self
            .iter()
            .map(|value| T::to_dxr(value))
            .collect::<Result<Vec<Value>, DxrError>>();

        Ok(Value::array(Array::new(values?)))
    }
}

impl<T> ToDXR for HashMap<String, T>
where
    T: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        let members = self
            .iter()
            .map(|(k, v)| T::to_dxr(v).map(|v| Member::new(k.to_owned(), v)))
            .collect::<Result<Vec<Member>, DxrError>>();

        Ok(Value::structure(Struct::new(members?)))
    }
}

impl<T> ToDXR for HashMap<&str, T>
where
    T: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        let members = self
            .iter()
            .map(|(k, v)| T::to_dxr(v).map(|v| Member::new((*k).to_owned(), v)))
            .collect::<Result<Vec<Member>, DxrError>>();

        Ok(Value::structure(Struct::new(members?)))
    }
}

impl<T> ToDXR for (T,)
where
    T: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::array(Array::new(tuple_to_values_1(self)?)))
    }
}

impl<A, B> ToDXR for (A, B)
where
    A: ToDXR,
    B: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::array(Array::new(tuple_to_values_2(self)?)))
    }
}

impl<A, B, C> ToDXR for (A, B, C)
where
    A: ToDXR,
    B: ToDXR,
    C: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::array(Array::new(tuple_to_values_3(self)?)))
    }
}

impl<A, B, C, D> ToDXR for (A, B, C, D)
where
    A: ToDXR,
    B: ToDXR,
    C: ToDXR,
    D: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::array(Array::new(tuple_to_values_4(self)?)))
    }
}

impl<A, B, C, D, E> ToDXR for (A, B, C, D, E)
where
    A: ToDXR,
    B: ToDXR,
    C: ToDXR,
    D: ToDXR,
    E: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::array(Array::new(tuple_to_values_5(self)?)))
    }
}

impl<A, B, C, D, E, F> ToDXR for (A, B, C, D, E, F)
where
    A: ToDXR,
    B: ToDXR,
    C: ToDXR,
    D: ToDXR,
    E: ToDXR,
    F: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::array(Array::new(tuple_to_values_6(self)?)))
    }
}

impl<A, B, C, D, E, F, G> ToDXR for (A, B, C, D, E, F, G)
where
    A: ToDXR,
    B: ToDXR,
    C: ToDXR,
    D: ToDXR,
    E: ToDXR,
    F: ToDXR,
    G: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::array(Array::new(tuple_to_values_7(self)?)))
    }
}

impl<A, B, C, D, E, F, G, H> ToDXR for (A, B, C, D, E, F, G, H)
where
    A: ToDXR,
    B: ToDXR,
    C: ToDXR,
    D: ToDXR,
    E: ToDXR,
    F: ToDXR,
    G: ToDXR,
    H: ToDXR,
{
    fn to_dxr(&self) -> Result<Value, DxrError> {
        Ok(Value::array(Array::new(tuple_to_values_8(self)?)))
    }
}

// if needed, implementations for more arguments can be implemented
