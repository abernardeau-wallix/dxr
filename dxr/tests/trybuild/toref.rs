use dxr::FromDXR;

#[derive(FromDXR)]
pub struct Reference<'a> {
    string: &'a str,
}

fn main() {}
