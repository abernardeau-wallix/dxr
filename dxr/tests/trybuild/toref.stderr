error: proc-macro derive panicked
 --> tests/trybuild/toref.rs:3:10
  |
3 | #[derive(FromDXR)]
  |          ^^^^^^^
  |
  = help: message: Deriving FromDXR is not possible for structs that contain reference types. Try using a std::borrow::Cow for field "string" instead.
